examples/cosmos/ics20/bank.qnt

```diff
@@ -0,0 +1,240 @@
+// -*- mode: Bluespec; -*-
+
+/**
+ * A bank module for ICS-020 (Fungible Token Transfer).
+ *
+ * ICS-020 assumes a high-level "bank module" capable of minting, burning, and
+ * transfering tokens. This specification is held as general as possible,
+ * providing only the assumptions required by ICS-020.
+ *
+ * For a possible implementation, consider the Cosmos SDK bank module:
+ * https://docs.cosmos.network/v0.47/modules/bank/
+ *
+ * Thomas Pani, Informal Systems, 2023
+ */
+
+module bank {
+
+  /* ***************************************************************************
+   * TYPES
+   * **************************************************************************/
+
+  type Address = str
+  type Denomination = str
+  type Amount = int
+
+  type TokenBalances = Denomination -> Amount
+  type Accounts = Address -> TokenBalances
+
+  /* ***************************************************************************
+   * FUNCTIONAL LAYER, manipulating balances
+   * **************************************************************************/
+
+  // Get balances of all denominations in the given account `addr`.
+  pure def getBalances(accounts: Accounts, addr: Address) : TokenBalances = {
+    if (accounts.keys().contains(addr)) {
+      accounts.get(addr)
+    } else {
+      Map()
+    }
+  }
+
+  // Get the amount of tokens of denomination `denom` in `balances`.
+  pure def getBalance(balances: TokenBalances, denom: Denomination): Amount = {
+    if (balances.keys().contains(denom)) balances.get(denom) else 0
+  }
+
+  /* Return `accounts`, where `address` is updated to hold
+   * `currentAmount + amount` tokens of denomination `denom`.
+   * 
+   * `currentAmount` refers to the number of tokens of denomination `denom` held
+   * by `address` in `accounts`.
+   *
+   * If there currently is no account under `address`, return accounts including
+   * `address` with exactly `amount`-many `denom` tokens.
+   */
+  pure def manipulateAccount(accounts: Accounts, address: Address,
```

Comment:
```suggestion
  pure def changeBalance(accounts: Accounts, address: Address,
```

Would not that be a better name? Although the operator also adds a new account, this behavior can be understood as adding `amount` to 0. (e.g., all maps are initialized with 0 by default in Solidity, so there are precedents for this behavior).

Reply:
Renamed to `addToBalance`, I think that's most accurate